### PR TITLE
[plinth][plc][api] Draft implementation of 'modularExponentiation' builtin

### PR DESCRIPTION
**Pull request by bezirg**
This is a draft implementation. 
Supporting GHC8.10 for this builtin needs extra work; so we discussed with @zliu41 and postpone it for a bit later.
So as not to make the whole CI not build anymore with GHC8.10, the implementation of this builtin always fails for GHC8.10. The builtin is not enabled to run yet on the chain.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/plutus-core/src/PlutusCore/Numeric/Extras.hs
Comment (bezirg) [inline plutus-core/plutus-core/src/PlutusCore/Numeric/Extras.hs:1]: I couldn't find a better module name to place the implementation . Any suggestions? I could instead place this inside the Default/Builtins.hs but then I would litter that file with MagicHash,UnboxedSums and CPP

**Review by effectfully — APPROVED** [review]
Seems good.

File: plutus-core/plutus-core/src/PlutusCore/Numeric/Extras.hs
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Numeric/Extras.hs:1]: Most definitely not `Default/Builtins.hs`. `PlutusCore.Crypto.Utils`? I guess `PlutusCore.Numeric.Extras` is fine too.

File: plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifdef VERSION_ghc_bignum
+-- a.ka. ghc>=9 , a.k.a ghc-bignum added. Another way is to do MIN_VERSION_GLASGOW_HASKELL
+import GHC.Num.Integer
+
+modularExponentiation :: Integer -> Integer -> Natural -> BuiltinResult Natural
+modularExponentiation b e m =
+    case integerPowMod# b e m of
+        (# n | #)  -> pure n
+        (# | () #) -> fail "modularExponentiation: failure"
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs:20]: @effectfully is it possible for a builtin function to return `Maybe`, or only builtin types can be returned?

File: stubs/plutus-ghc-stub/src/Plugins.hs [hunk 2]
```diff
@@ -110,9 +109,7 @@ instance Semigroup PluginRecompile where
 
 instance Monoid PluginRecompile where
   mempty = NoForceRecompile
-#if __GLASGOW_HASKELL__ < 840
```
Comment (zliu41) [inline stubs/plutus-ghc-stub/src/Plugins.hs:113]: Can you remind me what the GHC stub stuff is for?

**Review by zliu41 — COMMENTED** [review]
Looks reasonable. Regarding costing, make sure to follow the process outlined in https://github.com/IntersectMBO/plutus/pull/5874.

File: plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+{-# LANGUAGE UnboxedSums #-}
+module PlutusCore.Crypto.ModExp
+    ( modularExponentiation
+    ) where
+
+import PlutusCore.Builtin
+
+import GHC.Natural
+
+#ifdef VERSION_ghc_bignum
```
Comment (kwxm) [inline plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs:12]: This doesn't seem to work.  I've got GHC 9.6.6 but if I try to run any script involving `modularExponentiation` I get errors like this:
```
An error has occurred:
The machine terminated because of an error, either from a built-in function or from an explicit use of 'error'.
Caused by: modularExponentiation 9 2 17
modularExponentiation: FIXME: stub for GHC8.10, report to plutus developers
```
If I change `ifdef` to `ifndef` then it works.

File: plutus-tx/src/PlutusTx/Builtins/Internal.hs [hunk 4]
```diff
@@ -822,3 +823,16 @@ replicateByte n w8 =
       Haskell.error "byteStringReplicate errored."
     BuiltinSuccess bs -> BuiltinByteString bs
     BuiltinSuccessWithLogs logs bs -> traceAll logs $ BuiltinByteString bs
+
+{-# NOINLINE modularExponentiation #-}
```
Comment (kwxm) [inline plutus-tx/src/PlutusTx/Builtins/Internal.hs:827]: I think this should probably be exported from [PlutusTx.Builtins](https://github.com/IntersectMBO/plutus/blob/f74023ec043bd28d4fd16008f61b8d5d138cc73f/plutus-tx/src/PlutusTx/Builtins.hs#L128) as well (I've been caught out by that before).  Most of the other builtins have to have a wrapper to handle the `Builtin*` types in that file as well, but I think you can get away without that if everything is an `Integer` (maybe see [this comment](https://github.com/IntersectMBO/plutus/blob/f74023ec043bd28d4fd16008f61b8d5d138cc73f/plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs#L99)?).

Comment (kwxm) [reply plutus-tx/src/PlutusTx/Builtins/Internal.hs:827]: Maybe it also has to go in [PlutusTx.Prelude](https://github.com/IntersectMBO/plutus/blob/master/plutus-tx/src/PlutusTx/Prelude.hs).  I'm always confused by this.

**Review by kwxm — COMMENTED** [review]
This looks OK, although I'd be inclined to get rid of `Natural` and maybe give the function a less cumbersome name.

File: plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs:20]: Depends on what we want `Nothing` to mean. If we want it to fail evaluation, then we can do that. If we want `Nothing` to become some kind of UPLC value, then no, unless we add `Maybe` to the universe or treat `[]` as `Nothing` (and `[a]` as `Just x`) or teach builtins about SOPs or...

I'd be against making `Nothing` mean failure, not only because it'd be weird to treat a common Haskell value as a UPLC exception, but also because every failure must come with an error message and `Maybe` doesn't have that functionality. So yeah, wanna fail a builtin -- you have to use `BuiltinResult` as your monad.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs [hunk 5]
```diff
@@ -31,6 +31,7 @@ import Data.Word
 import GHC.Exts (Int (I#))
 import GHC.Integer
 import GHC.Integer.Logarithms
+import GHC.Natural
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: I think adding and maintaining a bunch of code for `Natural` is perfectly reasonable and I wanted to do that long before this PR, I just didn't care enough about it. That being said, after reading Kenneth's reasoning I agree that we should probably pass the `Integer` in and check that it's `>= 1` at least. Why `>= 2` BTW? `mod 1` seems to be uncontroversial?

[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > Why >= 2 BTW? mod 1 seems to be uncontroversial?

Well `mod 1` is identically zero (if you divide something by 1 then you have a remainder of 0), so it's kind of pointless. It seems like the kind of thing where an implementer might choose to raise an error or something instead, so you might get unexpected behaviour.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs [hunk 6]
```diff
@@ -221,6 +222,11 @@ instance ExMemoryUsage Integer where
     memoryUsage i = singletonRose $ memoryUsageInteger i
     {-# INLINE memoryUsage #-}
 
+instance ExMemoryUsage Natural where
+    -- FIXME: this is an over-approximation. Figure it out later when costing is done.
```
Comment (kwxm) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:226]: If you're going to do this then it should probably be the same as for `Integer`.  It would be a bit weird if the memory usage of `1234567::Natural` and `1234567::Integer` were different.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > Well mod 1 is identically zero (if you divide something by 1 then you have a remainder of 0), so it's kind of pointless.

Sure, but somebody may have some general code and they may rely on `x mod 1` being `0`. It's like saying that `Semigroup` for `()` is kind of pointless: sure it is, but you still want to have it in case it gets triggered in some general code, except in case of `mod` it's much more important, since you'll get a runtime error for some perfectly reasonable code, not even a type one.

So I'd be much in favor of doing reasonable things regardless of how useless they are, because somebody may rely on reasonable things working.

> I'm scared that we might end up with a natural type in PLC and then we'd have to duplicate all of the arithmetic and comparison functions...

No, I don't think that would make any sense and I'd be against doing it. However, `Natural` does fit very well in the `Universe` module where we also have all the `IntN` and `WordN` types, which we don't add to the universe either.

I'm not saying we _should_ add handling of `Natural` to the builtins machinery, let alone in this PR (as I agree with your reasoning on the `0` edge case), I'm only saying that I don't think that would be problematic or have high maintenance cost.

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > I'm not saying we should add handling of Natural to the builtins machinery, let alone in this PR (as I agree with your reasoning on the 0 edge case), 

Oh, stupid me, we've already done in a previous PR. Yeah, anyway, I think it's fine and removing that code would be fine as well.

**Review by effectfully — COMMENTED** [review]
I'm going to withdraw my approval after reading Kenneth's comments, I think he has great points. I don't disapprove the PR either though.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > > 
> 
> Sure, but somebody may have some general code and they may rely on `x mod 1` being `0`.

I'd be willing to be that that'll never happen, and if it does then someone's made a mistake.  This could be one of these endless pointless arguments like the one about what 0^0 is though.  I will point out that it also makes sense to define arithmetic modulo negative numbers (`ℤ_{-n}` is exactly the same as `ℤ_n`) but we're not proposing to allow that either.

File: stubs/plutus-ghc-stub/src/Plugins.hs (hunk 2 above)
Comment (bezirg) [reply stubs/plutus-ghc-stub/src/Plugins.hs:113]: I also dnt know what the ghc stub is for.

I did this change since the number `840` points to ghc `8.40.X` which is not a ghc version ever released. Most likely the original author wanted to say `804` which is ghc `8.4.X`. Further, since we don't even support ghc<8.4 , I thought the check is redundant

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > I'm actually not sure why. Is that because GHC simply doesn't export such a primitive?
Yes. The [integerPowMod# function](https://hackage.haskell.org/package/ghc-bignum-1.3/docs/src/GHC.Num.Integer.html#integerPowMod) takes a `Natural` as the final argument (the modulus), which is why we're talking about naturals in the first place.

File: plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs [hunk 7]
```diff
@@ -167,6 +168,7 @@ data DefaultFun
     | RotateByteString
     | CountSetBits
     | FindFirstSetBit
+    | ModularExponentiation
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:171]: Ok I go with PowModInteger.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > I don't know, to me it seems like banning n `div` 1 because it's trivial. 

It's not so much that  it's trivial, it's that I'm worried that it might require some special treatment (there's a special case for `naturalIsOne m` in the implementation of `integerPowMod`, for example), and since it's a bit non-standard there's  a possibility that if we change to some other implementation at some point in the future then the behaviour of that case might change.  I'm just a bit worried about stability.  If we do allow modulus 1 then I think we should add a test that you always get zero (and not a failure, say).

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: > (there's a special case for `naturalIsOne m` in the implementation of `integerPowMod`, for example)

There are plenty of special cases in there, if we count the base and the power.

> since it's a bit non-standard there's a possibility that if we change to some other implementation at some point in the future then the behaviour of that case might change. I'm just a bit worried about stability. If we do allow modulus 1 then I think we should add a test that you always get zero (and not a failure, say).

Yeah, a simple test will ensure that the behavior gets preserved.

Anyways, these are just my two cents, feel free to ignore.

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: @kwxm Can you summarize your objections?

From what I gather:

1) Natural should not be handled in the universe  (`AsInteger`), but should be handled manually in the implementation by checking that the arg(s) is/are `>=0`

2) We should not allow `mod=1` because it is trivial.

Some other objection?

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: @kwxm Answering to 1: 

I am following the pattern that @effectfully has laid out, in that if and arg/result is clearly a super-type of Integer (Word,Word8,16,32,64,Int,Int8,16,32,64) we handle it in the Universe. For the example you mentioned:

```
-- https://github.com/IntersectMBO/plutus/blob/f74023ec043bd28d4fd16008f61b8d5d138cc73f/plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs#L1269

indexByteStringDenotation :: BS.ByteString -> Int -> BuiltinResult Word8
            indexByteStringDenotation xs n = do
                unless (n >= 0 && n < BS.length xs) $ ...
```

As you can see, we do utilize the pattern I am mentioned in the result type. We don't upcast the result to Integer (although we only support Integers). The universe will do that via the `AsInteger` typeclass.

I would actually argue in the above implementation that we could strip off the `n>=0` check and put `Word` in there. 

I like this pattern because we do these conversions uniformly for all our implementations (less code, and less likely to lead to errors).

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: @kwxm I do not agree about (2) : not allowing mod=1 because it is trivial, similar to what @effectfully  said that we allow `div 1` and `mod 1`

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: About what is the proper mathematical definition of `modular exponentiation`, I am not a mathematician so I am following what was described in CIP-109: https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0109/ . Maybe @perturbing can comment about what mathematical definition we actually want here.

I can only comment about the specific interfaces/implementations we have in hand.

There are 2 fast interfaces in Haskell (labelled here as OLD and NEW) that provide *some kind of* "modular exponentation" operation.
Both of these interfaces utilise underneath the gmp library.

# OLD (powModInteger)

`powModInteger :: Integer -> Integer -> Integer -> Integer`

As you can see this has everywhere Integers, which does not make complete sense for the 3rd argument and the result;
it would be better to type them as Naturals.

Keeping the same interface, there are 2 different behaviors of the `powModInteger`

Behavior1 of powModInteger with GHC<=8.10:

```
-- mod: 0
ghci810> powModInteger 1 1 0
ghc: libraries/integer-gmp/cbits/wrappers.c:671: integer_gmp_powm1: Assertion `m0' failed.
Aborted (core dumped)

-- negative exponent with non-prime mod
ghci> powModInteger 2 (-5) 4
Floating point exception (core dumped)
```

As you can see this powModInteger+<=ghc8.10 is very dangerous: it fails catastrophically **without a way**
to catch the exception.
If we want to use this interface, we would then need to validate before-hand the
input ourselves before calling the function. Specifically, *I think* we would need
to validate that the mod arg is a prime number (which further implies >=0).

Behavior2 of powModInteger with GHC>=9:

```
-- mod: 0
ghci810> powModInteger 1 1 0
In the use of ‘powModInteger’
(imported from GHC.Integer.GMP.Internals):
Deprecated: "Use integerPowMod# instead"
0

-- negative exponent with non-prime mod
ghci> powModInteger 2 (-5) 4
In the use of ‘powModInteger’
(imported from GHC.Integer.GMP.Internals):
Deprecated: "Use integerPowMod# instead"
0
```

Well, how come this work? Well, this powModInteger (deprecated)
has changed implementations and points
now underneath to the new implementation (the implementation used also by the NEW interface). Instead of failing
catastrophically, it instead maps all the
those values that don't have a result arbritrarily to 0:

```
{-# DEPRECATED powModInteger "Use integerPowMod# instead" #-}
powModInteger :: Integer -> Integer -> Integer -> Integer
powModInteger b e m = case I.integerPowMod# b e (I.integerToNatural m) of
   (# r  | #) -> I.integerFromNatural r
   (# | () #) -> 0
```

# NEW (integerPowMod#)

Is provided only for GHC>=9.

```
integerPowMod# :: Integer -> Integer -> Natural -> (# Natural | () #)
```

As you can see it types the 3rd mod-arg and the result better.
Second, it never fails catastrophically, even when `mod-arg == 0 :: Natural`.

```
ghci9> integerPowMod# 1 1 0
(# | () #)
ghci9> integerPowMod 2 (-5) 4
(# | () #)


# What about Cryptonite's `expFast` function?

It is just an alias to the OLD `powModInteger` interface described above.
Which means that cryptonite+GHC<=8.10 fails catastrophically,
whereas cryptonite+GHC>=9 returns 0.

To complicate things, cryptonite will fallback to a pure-Haskell implementation
when it cannot find gmp lib. Which might lead to 2 different behaviors
even under the same GHC version:

```
-- ghc9,WITH-gmp,cryptonite-0.30
ghci9> expSafe 2 (-5) 4
0
-- ghc9,WITHOUT-gmp,cryptonite-0.30
>exponentiation 2 (-5) 4
*** Exception: stack overflow
```
# What about Cryptonite's `expSafe` function?

This is an adaptation that avoids a side-channel attack on the gmp implementation.
I don't think we need it, but if you want to know more about it / discuss it , we can do that.


# TLDR:

IMO: powModInteger is unusable, cryptonite is messy, and the only sane choice
is integerPowMod# , which unfortunately requires GHC>=9.

My idea was to kind of manually "backport" what `integerPowMod#` does to GHC8.10
iff deemed necessary, since we might drop support of GHC8.10 soon.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 6 above)
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:226]: You are right, in hindsight it should be the same since Natural goes via Integer in our case.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: Furthermore, I don't know what should be the proper result when mod-arg is 0, but for reference, here is the behaviour of modular exponentation in python3's stdlib:

```
$python
Python 3.12.4 (main, Jun  7 2024, 06:33:07) [GCC 14.1.1 20240522] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> pow(1,1,0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: pow() 3rd argument cannot be 0
>>> pow(2,-5,4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: base is not invertible for the given modulus
>>>>
```

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: CIP-109  https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0109/ says about this:

> $b$ is an integer, the exponent $e$ a non-negative integer and the modulus $m$ a positive integer.
> [...]
> We propose to also include this extension [negative exponent e ]to the plutus built-in, for optimized inversion when this is possible. This is inversion is not guaranteed to exist for all numbers $b$, only when the modulus $m$ is a prime number this is guaranteed. 

> We also propose that the built-in fails if this inverse does not exist, and if a modulus is provided that is smaller than one.

So from what I understand from the CIP-109, the interface of the builtin looks like this:

- base : integer
- exponent : integer
- modulo : integer
- result: integer

If the modulo is <= 0  , fail.
If the exponent is < 0 , *SOMETIMES* fail.

So I think the above expected results are matched by `integerPowMod#`.

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:34]: @kwxm @effectfully 

After discussing with @effectfully , I should have made it more explicit:

`integerPowMod#` returns `()` when the mod-arg is 0
=>
 which makes this PR's builtin implementation to `fail`, as expected by CIP-109.

No `nonZeroThirdArg` --- same in spirit as `nonZeroSecondArg` ---  is needed.

**Review by effectfully — APPROVED** [review]
Nikos very helpfully explained to me that the builtin already behaves like I want it to. Conversion of the last argument to `Natural` filters out negative integers and `integerPowMod#` returns an explicit failure when the last argument is `0`. So by my standards it's all good, I'd only like to see unit tests demonstrating all this behavior.

@bezirg could you please add unit tests to `Evaluation.Builtins.Definition` similar to all the other unit tests that we have there (like those in `test_Integer`, `test_Crypto` etc)?

**Comment by kwxm** [comment]
> Nikos very helpfully explained to me that the builtin already behaves like I want it to. Conversion of the last argument to Natural filters out negative integers and integerPowMod# returns an explicit failure when the last argument is 0. So by my standards it's all good, I'd only like to see unit tests demonstrating all this behavior.

I'd prefer to put an explicit check for modulus > 0 in the builtin.  `integerPowMod#` does this itself (although you only discover that by looking at it's source code), but let's be sure in case something changes.  I've looked at a number of implementations in different languages and libraries and so far I've found three different behaviours when the modulus is zero.  There's no commonly-accepted agreement on what the behaviour should be, and none of the things I've looked at actually document what the behaviour is: you have to use trial and error or examine the code to find out.  This is really fragile, so let's just ban it outright!  I think we should accept m=1 always giving you a^b mod m = 0 since that seems to be pretty universal , but we should still add a test. 

I'm still a bit sceptical about using `Natural` to enforce bounds for us. We already have multiple builtins that do their own bound checks, but now we have two ways of checking n>=0, which could be confusing. Maybe we should use `Natural` everywhere we do that though.

**Comment by effectfully** [comment]
> I'd prefer to put an explicit check for modulus > 0 in the builtin. 

Just to be clear, I don't mind that at all. Only saying that for me both the options work, so if you feel like we should really make it an explicit check, we can invoke the "whoever has the strongest opinion wins" clause.

> This is really fragile

I don't think it is! We should cover all edge cases with tests. We can even make both unit and golden tests just to be dead sure that the behavior isn't gonna change. And I don't feel like making the check explicit will contribute much towards stability of behavior, given proper testing.

>  I think we should accept m=1 always giving you a^b mod m = 0 since that seems to be pretty universal , but we should still add a test.

:+1: 

> I'm still a bit sceptical about using Natural to enforce bounds for us. We already have multiple builtins that do their own bound checks, but now we have two ways of checking n>=0, which could be confusing. Maybe we should use Natural everywhere we do that though.

The underlying primitive `base` function has `Natural` in its type, so we just conveniently wrap it directly instead of checking the bounds manually and converting the `Integer` argument manually to `Natural` in order to pass it to the function. Less boilerplate and it's still pretty clear what's going on, since we do the same for `Int64`, `Word8` etc etc. Again, no strong inclinations from me, but I feel like using `Natural` is pretty reasonable and follows existing conventions.

I do agree that there are two ways to check bounds (explicitly and implicitly) and it may be confusing, but both are used extensively, are we going to implement bounds checking manually for all the various integral types just so that all bounds checking is done explicitly?

File: plutus-core/plutus-core/src/PlutusCore/Crypto/ExpMod.hs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        (# | () #) -> fail "expMod: failure"
+#else
+-- FIXME: fugly stub implementation to make the various test-suites/CI pass for GHC8.10.
+-- This means that we cannot provide random testing for expMod at the moment.
+expMod _ _ 0 = fail "Cannot divide by zero"
+expMod 500 0 500 = pure 1
+expMod 500 5 500 = pure 0
+expMod 1 (-3) 4 = pure 1
+expMod 2 (-3) 3 = pure 2
+expMod 4 (-5) 9 = pure 4
```
Comment (bezirg) [inline plutus-core/plutus-core/src/PlutusCore/Crypto/ExpMod.hs:31]: Super ugly hack to make things happy for GHC8.10 , without disabling building for ghc8.10 for now.

**Comment by bezirg** [comment]
I added some unit tests. I didn't add random testing of the builtin because of the stub code i have for ghc8.10

File: plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Definition.hs [hunk 9]
```diff
@@ -573,6 +574,27 @@ test_Integer = testNestedM "Integer" $ do
     evals False LessThanEqualsInteger [] [cons @Integer 4001, cons @Integer 4000]
     evals True EqualsInteger [] [cons @Integer (-101), cons @Integer (-101)]
     evals False EqualsInteger [] [cons @Integer 0, cons @Integer 1]
+    for_ [DivideInteger, QuotientInteger, ModInteger, RemainderInteger] $ \ b ->
+        fails (lowerInitialChar $ show b <> "-div-by-zero") b [] [cons @Integer 1, cons @Integer 0]
+    test_ExpModInteger
+
+test_ExpModInteger :: TestNested
+test_ExpModInteger = testNestedM "ExpMod" $ do
```
Comment (kwxm) [inline plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Definition.hs:582]: I'm not objecting to these tests (more tests are better!) but maybe there's a discussion to be had about whether we need unit tests like this (which we have for lots of other builtins as well) when we also have the conformance tests.  I suppose these get run more often than the conformance tests though.
